### PR TITLE
jslt-152: parse-url error handling control

### PR DESCRIPTION
**Pull request by mrange**
See: https://github.com/schibsted/jslt/issues/152

### CODE REVIEW

File: core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java [hunk 1]
```diff
@@ -1303,8 +1305,17 @@ public JsonNode call(JsonNode input, JsonNode[] arguments) {
         if(aURL.getUserInfo() != null && !aURL.getUserInfo().isEmpty())
           objectNode.put("userinfo", aURL.getUserInfo());
         return objectNode;
-      } catch (MalformedURLException | UnsupportedEncodingException e) {
-        throw new JsltException("Can't parse " + urlString, e);
+      } catch (Throwable e) {
```
Comment (mrange) [inline core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java:1308]: Possible regression as this catch is different. However, we ran into `InvalidArgumentException` when having trailing `%` which with the current code pattern gave no hints on the problem string.

File: core/src/test/resources/query-tests.yaml [hunk 2]
```diff
@@ -460,3 +460,36 @@ tests:
...
+# parse-url tests with throwOnFailure set to false should return an error object for bad cases
+
+  - input:
+    query: parse-url("this-is-an-invalid-url", false)
+    output: |
+      {
+        "error": "MalformedURLException",
+        "message": "no protocol: this-is-an-invalid-url",
+        "input": "this-is-an-invalid-url"
+      }
```
Comment (larsga) [inline core/src/test/resources/query-tests.yaml:495]: We need a test of what happens if the second argument is not a boolean. IMHO that should be considered an error: ie, throw an exception.

We also need a test of `null` handling, with and without the second parameter. (Should have had a `null` test already, sorry.) I think `null` input should lead to `null` output, regardless of the value of the second parameter.

File: functions.md [hunk 3]
```diff
@@ -826,10 +826,13 @@ format-time(null, "yyyy-MM-dd")                  => null
...
 ## Miscellaneous functions
 
-### _parse-url(url) -> object_
+### _parse-url(url, throwOnFailure?) -> object_
 
 Parses `url` and returns an object with keys [`scheme`, `userinfo`, `host`, `port` `path`, `query`, `parameters`, `fragment` ]
 
+If the optional `throwOnFailure` argument is not specified invalid URLs will generate an exception. If `throwOnFailure` is set to `false` the `parse-url` returns
+an object indicating an error occurred.
+
```
Comment (larsga) [inline functions.md:835]: We need to document the shape of that object, and have an example.

File: core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java [hunk 4]
```diff
@@ -1303,8 +1305,17 @@ public JsonNode call(JsonNode input, JsonNode[] arguments) {
...
+        if (throwOnFailure) {
+          throw new JsltException("Can't parse " + urlString, e);
+        } else {
+          final ObjectNode errorNode = NodeUtils.mapper.createObjectNode();
+          Class errorClass = e.getClass();
+          errorNode.put("error", errorClass.getSimpleName());
```
Comment (larsga) [inline core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java:1314]: Why not the full class name?

File: core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java (hunk 1 above)
Comment (larsga) [reply core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java:1308]: `Throwable` is perhaps overdoing it, since it's going to catch `OutOfMemoryError` and similar things. Perhaps just `Exception`?

File: core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java (hunk 4 above)
Comment (mrange) [reply core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java:1314]: I don't feel strongly over this. I can change to full class name.

File: functions.md (hunk 3 above)
Comment (mrange) [reply functions.md:835]: I added some examples below but I agree we could add more documentation on the shape both when good and bad input

File: core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java (hunk 1 above)
Comment (ole-magnus) [reply core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java:1308]: `Throwable` -> `Exception` makes sense 👍  We just rushed the fix out without considering that.

Comment (jarlebass) [reply core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java:1308]: Or we can simulate [`NonFatal`](https://github.com/scala/scala/blob/cbe6ade1744bb316a63a07b07d3bf448590ad2cd/src/library/scala/util/control/NonFatal.scala#L41) :)

Comment (mrange) [reply core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java]: Now is `Exception`

File: functions.md (hunk 3 above)
Comment (mrange) [reply functions.md:835]: Updated with more detaild examples.

File: core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java (hunk 4 above)
Comment (mrange) [reply core/src/main/java/com/schibsted/spt/data/jslt/impl/BuiltinFunctions.java]: Changed to use `getName`

File: core/src/test/resources/query-tests.yaml (hunk 2 above)
Comment (mrange) [reply core/src/test/resources/query-tests.yaml:495]: I am unsure how you typically tests for failures. Do you have an example? I looked in the query-tests.yaml and found no obvious cases to me

Comment (larsga) [reply core/src/test/resources/query-tests.yaml:495]: See `function-error-tests.json` for examples.

Comment (mrange) [reply core/src/test/resources/query-tests.yaml:501]: Added tests

**Comment by mrange** [comment]
I updated the code as I interpreted the feedback.

**Comment by biochimia** [comment]
I would like to understand a little bit more the reasoning for using a `throwOnFailures` parameter, as opposed to a fallback parameter as is customary (as pointed out in #152).

My concern is that the proposed approach is not general and won't cleanly apply to cases where the return value is not an object. Is `parse-url` a special enough case that it warrants its own error handling mechanism?

I suppose we're bound to find other cases where exception handling is warranted, and having a common way to handle errors would be preferable. Say a `try ... catch ...` expression could be introduced in the language.

**Comment by mrange** [comment]
This "fix" was rushed as we had a pressing concern to address in our pipeline so I don't claim it's perfect as is.

When it comes to `parse-url` we are interested in why it fails. With a fallback param that information is lost. `parse-url` as of today already returns an object so in this particular case I don't see any practical issues but I agree there's a break in the idioms used today.

I am personally not so much a favour of exceptions as no one expects the unexpected exception them ;)
